src/signature.rs

```diff
@@ -307,11 +339,51 @@ fn build_hostname(service: &str, region: &Region) -> String {
 	}
 }
 
+/// extract_s3_redirect_location takes a Hyper Response and attempts to pull out the temporary endpoint.
+fn extract_s3_redirect_location(response: Response) -> Result<String, AWSError> {
+	// Double checking this feels like belts and suspenders since we're checking the status code
+	// before calling this.  Remove this check?
+
+	// Verify it's a 307 temporary redirect
+	if response.status != HTTP_TEMPORARY_REDIRECT {
+		return Err(AWSError::new("Trying to find temporary location when status is not 307 temp redirect."))
+	}
+
+	let mut reader = EventReader::new(response);
+	let mut stack = XmlResponseFromAws::new(reader.events().peekable());
+	stack.next(); // xml start tag
+
+	// extract and return temporary endpoint location
+	extract_s3_temporary_endpoint_from_xml(&mut stack)
+}
+
+/// extract_s3_temporary_endpoint_from_xml takes in XML and tries to find the value of the Endpoint node.
+fn extract_s3_temporary_endpoint_from_xml<'a, T: Peek + Next>(stack: &mut T) -> Result<String, AWSError> {
+	try!(start_element(&"Error".to_string(), stack));
+
+	// now find Endpoint contents
+	// This may infinite loop if there's no endpoint in the response: how can we prevent that?
+	loop {
+		let current_name = try!(peek_at_name(stack));
+		if current_name == "Endpoint" {
+			let obj = try!(string_field("Endpoint", stack));
+			return Ok(obj);
+		}
+		stack.next();
+	}
+	Err(AWSError::new("Couldn't find redirect location for S3 bucket"))
```

Comment:
Compiler warning on this, we should figure out how to ensure LOOP OF INFINITY +1 doesn't happen.